include/xtensor/xrandom.hpp

```diff
@@ -0,0 +1,195 @@
+/***************************************************************************
+* Copyright (c) 2016, Johan Mabille and Sylvain Corlay                     *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+/**
+ * @brief functions to obtain xgenerators generating random numbers with given shape
+ */
+
+#ifndef XRANDOM_HPP
+#define XRANDOM_HPP
+
+#include <utility>
+#include <random>
+#include <memory>
+#include <sstream>
+
+#include "xgenerator.hpp"
+
+namespace xt
+{
+    namespace detail
+    {
+        template <class D, class S, class V, class E = std::mt19937>
+        struct random_impl
+        {
+            using value_type = V;
+            using size_type = std::size_t;
+
+            random_impl(E engine, D dist, S shape) :
+                m_engine(engine), m_dist(dist), m_strides(shape.size()), 
+                m_engine_state(std::make_shared<std::stringstream>())
+            {
+                std::size_t data_size = 1;
+                for(size_type i = m_strides.size(); i != 0; --i)
+                {
+                    m_strides[i - 1] = data_size;
+                    data_size = m_strides[i - 1] * shape[i - 1];
+                }
+                // save initial engine state
+                (*m_engine_state) << m_engine; 
+            }
+
+            template <class... Args>
+            inline value_type operator()(Args... args) const
+            {
+                std::array<size_type, sizeof...(Args)> args_arr({static_cast<size_type>(args)...});
+                return access_impl(args_arr);
+            }
+
+            inline value_type operator[](const xindex& idx) const
+            {
+                return access_impl(idx);
+            }
+
+            template <class It>
+            inline value_type element(It first, It /*last*/) const
+            {
+                return access_impl(*first);
+            }
+
+        private:
+            mutable E m_engine;
+            mutable D m_dist;
```

Comment:
the distributions can have an internal state, so calling `operator()` can modify them.